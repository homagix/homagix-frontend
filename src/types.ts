export type User = {
  id: string
  firstName: string
}

export type AppError = {
  message: string
  details?: unknown
  link?: string
}

export type Item = {
  id: string
  amount: number
}

export type Dish = {
  id: string
  name: string
  image: string
  recipe?: string
  source?: string
  last?: string
  items: Item[]
  alwaysOnList: boolean
  isFavorite?: boolean
}

export type Ingredient = {
  id: string
  amount: number
  unit: string
  name: string
}
