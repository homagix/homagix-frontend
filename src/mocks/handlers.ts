import { rest } from "msw"

export const handlers = [
  rest.get("http://localhost:8200/api/dishes", (req, res, ctx) => {
    return res(ctx.json({ dishes: [{ id: 1, name: "Pizza" }]}))
  }),
]
