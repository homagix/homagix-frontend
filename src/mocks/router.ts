import { defineComponent } from "vue"
import { createWebHistory } from "vue-router"
import { createRouter } from "vue-router"

export function createRouterMock(component: ReturnType<typeof defineComponent>, extraPaths: string[] = []) {
  const history = createWebHistory()
  const routes = ["/", ...extraPaths].map(path => ({ path, component }))
  return createRouter({ history, routes })
}
