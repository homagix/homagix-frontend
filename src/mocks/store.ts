import { vi } from "vitest"
import { createStore } from "vuex"

export function createStoreMock(state: object = {}, options: { actions?: string[]; mutations?: string[] } = {}) {
  function createMocksRecord(names?: string[]) {
    if (!names) {
      return undefined
    }
    return Object.assign({}, ...names.map(name => ({ [name]: vi.fn() })))
  }

  const actionMocks = createMocksRecord(options.actions)
  const mutationMocks = createMocksRecord(options.mutations)
  const store = createStore({
    state() {
      return state
    },
    actions: actionMocks,
    mutations: mutationMocks,
  })
  return { store, actionMocks, mutationMocks }
}
