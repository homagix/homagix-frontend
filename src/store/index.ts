import { Store as VuexStore, useStore as baseUseStore, createStore, ActionContext } from "vuex"
import mutations from "@/store/mutations"
import actions from "@/store/actions"
import getters from "@/store/getters"
import { AppError, Dish, Ingredient } from "@/types"
import { state as auth } from "@/auth"

const state = {
  error: null as AppError | null,
  dishes: [] as Dish[],
  ingredients: [] as Ingredient[],
  ...auth,
}

const store = createStore({
  strict: import.meta.env.NODE_ENV !== "production",
  state,
  mutations,
  actions,
  getters,
})

function listKeys(obj: object) {
  return Object.assign({}, ...Object.keys(obj).map(key => ({ [key]: key })))
}

export default store
export type State = typeof state
export type Mutations = typeof mutations
export type Actions = typeof actions
export type Getters = typeof getters
export const MutationType = listKeys(mutations) as Record<keyof Mutations, keyof Mutations>
export const ActionType = listKeys(actions) as Record<keyof Actions, keyof Actions>

type MutationsProp = {
  commit<K extends keyof Mutations>(key: K, payload: Parameters<Mutations[K]>[1]): ReturnType<Mutations[K]>
}

export type Context = Omit<ActionContext<State, State>, "commit"> & MutationsProp
export type Store = VuexStore<State>

export function useStore(): Store {
  return baseUseStore()
}
