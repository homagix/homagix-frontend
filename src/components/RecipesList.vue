<script setup lang="ts">
import { computed } from "vue"
import { Dish } from "@/types"
import { useStore } from "@/store"
import { useRouter } from "vue-router"

const store = useStore()
const router = useRouter()

const sortedDishes = computed(() => {
  const filteredDishes = store.state.dishes?.filter(d => !d.alwaysOnList)
  return filteredDishes?.sort((d1, d2) => d1.name.localeCompare(d2.name)) as Dish[]
})

function activate(id: string) {
  router.push({ name: "recipe", params: { id } })
}
</script>

<template>
  <h2>Rezepte</h2>
  <ul id="recipes-list">
    <li v-for="dish in sortedDishes" :key="dish.id" @click="() => activate(dish.id)">
      <o-icon icon="utensils"></o-icon> {{ dish.name }}
    </li>
  </ul>
</template>

<style scoped lang="scss">
li {
  cursor: pointer;

  &:hover {
    background-color: #eee;
  }
}
</style>
