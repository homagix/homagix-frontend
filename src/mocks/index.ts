import { mount, shallowMount } from "@vue/test-utils"
import { createRouterMock } from "./router"
import { createStoreMock, StoreMockOptions } from "./store"
import Oruga from "@oruga-ui/oruga-next"

type MountOptions = {
  shallow?: boolean
  props?: Record<string, any>
  state?: object
  storeOptions?: StoreMockOptions
  extraPaths?: string[]
}

export function mountComponent<C>(component: C, options: MountOptions) {
  const { store, actionMocks, mutationMocks } = createStoreMock(options.state || {}, options.storeOptions || {})
  const router = createRouterMock(component, options.extraPaths || [])
  const m = options.shallow === undefined || options.shallow ? shallowMount : mount
  const wrapper = m(component, { props: options.props || {}, global: { plugins: [store, router, Oruga] } })

  return { wrapper, store, router, actionMocks, mutationMocks }
}
