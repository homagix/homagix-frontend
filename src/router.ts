import { createRouter, createWebHistory } from "vue-router"
import RecipesList from "@/components/RecipesList.vue"

export default createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    { name: "home", path: "/", redirect: "/recipes" },
    { name: "recipes", path: "/recipes", component: RecipesList },
  ],
})
