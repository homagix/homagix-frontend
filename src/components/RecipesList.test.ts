import { describe, expect, it, vi } from "vitest"
import { mount } from "@vue/test-utils"
import { createRouterMock } from "../mocks/router"
import { createStoreMock } from "../mocks/store"
import { Dish, Ingredient } from "@/types"
import Oruga from "@oruga-ui/oruga-next"
import RecipesList from "@/components/RecipesList.vue"

const dishes = [
  {
    id: "1",
    name: "dish 1",
    items: [],
    alwaysOnList: false,
    images: ["dish1.jpg"],
    isFavorite: false,
    isEditable: true,
  },
  {
    id: "2",
    name: "dish 2",
    items: [],
    alwaysOnList: false,
    images: ["dish2.jpg"],
    isFavorite: true,
    isEditable: true,
  },
] as Dish[]

const ingredients = [] as Ingredient[]

describe("RecipesList", () => {
  it("should render", () => {
    const router = createRouterMock(RecipesList, ["/wordcloud"])
    const { store } = createStoreMock({ dishes, ingredients }, { actions: ["SET_USER"] })
    const wrapper = mount(RecipesList, {
      global: { plugins: [store, router, Oruga] },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should render a message when there are no items in list", () => {
    const router = createRouterMock(RecipesList, ["/wordcloud"])
    const { store } = createStoreMock({ dishes: [], ingredients }, { actions: ["SET_USER"] })
    const wrapper = mount(RecipesList, {
      global: { plugins: [store, router, Oruga] },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should render favorites", () => {
    const router = createRouterMock(RecipesList, ["/wordcloud"])
    const { store } = createStoreMock({ dishes, ingredients }, { actions: ["SET_USER"] })
    const wrapper = mount(RecipesList, {
      global: { plugins: [store, router, Oruga] },
      props: { onlyFavorites: true },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
