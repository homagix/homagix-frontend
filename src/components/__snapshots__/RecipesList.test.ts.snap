// Vitest Snapshot v1

exports[`RecipesList > should render 1`] = `
"<div class=\\"title is-4\\" data-v-2a5a893a=\\"\\"> Rezepte
  <!--v-if--><a href=\\"/wordcloud\\" class=\\"\\" data-v-2a5a893a=\\"\\"><span class=\\"o-icon wordcloud\\" data-v-2a5a893a=\\"\\"><i class=\\"mdi mdi-magnifying-glass mdi-24px\\"></i></span></a>
</div>
<ul id=\\"recipes-list\\" data-v-2a5a893a=\\"\\">
  <li data-v-2a5a893a=\\"\\"><span class=\\"o-icon\\" data-v-2a5a893a=\\"\\"><i class=\\"mdi mdi-utensils mdi-24px\\"></i></span> dish 1 <a href=\\"#\\" class=\\"favorite\\" title=\\"Zur Favoritenliste hinzufügen\\" data-v-4723bbcc=\\"\\" data-v-2a5a893a=\\"\\">☆</a></li>
  <li data-v-2a5a893a=\\"\\"><span class=\\"o-icon\\" data-v-2a5a893a=\\"\\"><i class=\\"mdi mdi-utensils mdi-24px\\"></i></span> dish 2 <a href=\\"#\\" class=\\"favorite\\" title=\\"Zur Favoritenliste hinzufügen\\" data-v-4723bbcc=\\"\\" data-v-2a5a893a=\\"\\">☆</a></li>
</ul>"
`;
